### ISSUE TITLE
Hollow Coal Chimpanzee - Address.isContract() is not a reliable way of checking if the input is an EOA .

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-07-kyber-swap-judging#28]
Hollow Coal Chimpanzee

medium

# Address.isContract() is not a reliable way of checking if the input is an EOA .
## Summary
Address.isContract() is not a reliable way of checking if the input is an EOA .
## Vulnerability Detail
The `permit` function enables the owner of a token to authorize another address to perform ownership transfer on their behalf through a signature. The determination of whether the owner is a contract or an externally owned account (EOA) is made using OpenZeppelin's `Address.isContract()` function. However, this approach has been criticized for potential security vulnerabilities. This could lead to unexpected consequences as discussed in the following GitHub issue:  
https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3355.
This article also discusses the same issue:
https://despac1to.medium.com/carefully-use-openzeppelins-address-iscontract-msg-sender-4136cc6ff66d
```solidity
  function permit(
    address spender,
    uint256 tokenId,
    uint256 deadline,
    uint8 v,
    bytes32 r,
    bytes32 s
  ) external override onlyNotExpired(deadline) {
    bytes32 digest = keccak256(
      abi.encodePacked(
        '\x19\x01',
        DOMAIN_SEPARATOR,
        keccak256(
          abi.encode(PERMIT_TYPEHASH, spender, tokenId, _getAndIncrementNonce(tokenId), deadline)
        )
      )
    );
    address owner = ownerOf(tokenId);
    require(spender != owner, 'ERC721Permit: approval to current owner');


    if (Address.isContract(owner)) {
      require(
        IERC1271(owner).isValidSignature(digest, abi.encodePacked(r, s, v)) == 0x1626ba7e,
        'Unauthorized'
      );
    } else {
      address recoveredAddress = ecrecover(digest, v, r, s);
      require(recoveredAddress != address(0), 'Invalid signature');
      require(recoveredAddress == owner, 'Unauthorized');
    }


    _approve(spender, tokenId);
  }
```
## Impact
Relying on this determination may pose significant security risks and vulnerabilities.
## Code Snippet
https://github.com/sherlock-audit/2023-07-kyber-swap/blob/main/ks-elastic-sc/contracts/periphery/base/ERC721Permit.sol#L87
## Tool used

Manual Review

## Recommendation
In conclusion, for security considerations, it is not recommended to directly use the return value of Address.isContract() to determine whether a caller is a contract or not. require(msg.sender == tx.orign) works now and is a better practice.
